pkg/apiserver/storage/store.go

```diff
@@ -0,0 +1,464 @@
+	}
+	data, err := runtime.Encode(s.codec, obj)
+	if err != nil {
+		return err
+	}
+	objKey := newObjectKey(key)
+	dr := &v1alpha1.DataResource{
+		ObjectMeta: objKey.objectMeta(),
+		Data:       data,
+	}
+	// Set the UID of dataresource same as the resource stored in favor of preconditions on deleting
+	if metaobj, ok := obj.(metav1.Object); ok {
+		dr.UID = metaobj.GetUID()
+	}
+	ret, err := s.client.Create(dr)
+	if err != nil {
+		return err
+	}
+	if out != nil {
+		rv, err := s.versioner.ParseResourceVersion(ret.ResourceVersion)
+		if err != nil {
+			return err
+		}
+		err = decode(s.codec, s.versioner, ret.Data, out, int64(rv))
+		return err
+	}
+	return nil
+}
+
+func (s *store) Delete(ctx context.Context, key string, out runtime.Object, preconditions *storage.Preconditions) error {
+	objKey := newObjectKey(key)
+	return s.client.Delete(objKey.fullName(), &metav1.DeleteOptions{
+		Preconditions: &metav1.Preconditions{
+			UID: preconditions.UID,
+		},
+	})
+}
+
+func (s *store) Watch(ctx context.Context, key string, resourceVersion string, p storage.SelectionPredicate) (watch.Interface, error) {
+	// Client based watching hold a connection to apiserver for each watch request
+	// TODO: replace with a sharedInformer based watching strategy
+	return s.WatchList(ctx, key, resourceVersion, p)
+}
+
+func (s *store) WatchList(ctx context.Context, key string, resourceVersion string, p storage.SelectionPredicate) (watch.Interface, error) {
+	// Client based watching hold a connection to apiserver for each watch request
+	// TODO: replace with a sharedInformer based watching strategy
+	objKey := newObjectKey(key)
+	w, err := s.client.Watch(metav1.ListOptions{
+		LabelSelector:   objKey.labelSelectorStr(),
+		ResourceVersion: resourceVersion,
+	})
+	if err != nil {
+		return nil, err
+	}
+	outWatcher := newWatcherWrapperWithPrediction(s, w, p)
+	go outWatcher.run()
+	return outWatcher, nil
+}
+
+func (s *store) Get(ctx context.Context, key string, resourceVersion string, out runtime.Object, ignoreNotFound bool) error {
+	objKey := newObjectKey(key)
+	selector, err := objKey.selector()
+	if err != nil {
+		return err
+	}
+	ret, err := s.lister.List(selector)
+	if err != nil {
+		return err
+	}
+	if len(ret) == 0 {
+		if ignoreNotFound {
+			return runtime.SetZeroValue(out)
+		}
+		return storage.NewKeyNotFoundError(key, 0)
+	}
+	if len(ret) > 1 {
+		return storage.NewInternalError("more than 1 resources found")
+	}
+	rv, err := s.versioner.ParseResourceVersion(ret[0].ResourceVersion)
+	if err != nil {
+		return err
+	}
+	return decode(s.codec, s.versioner, ret[0].Data, out, int64(rv))
+}
+
+func (s *store) GetToList(ctx context.Context, key string, resourceVersion string, p storage.SelectionPredicate, listObj runtime.Object) error {
+	listPtr, err := meta.GetItemsPtr(listObj)
+	if err != nil {
+		return err
+	}
+	v, err := conversion.EnforcePtr(listPtr)
+	if err != nil || v.Kind() != reflect.Slice {
+		panic("need ptr to slice")
```

Comment:
return an error instead of just panic?

Reply:
This is the behavior of the official implementation, according to the context that this method get called, this condition indicates a bug in code and better to be treated as panic.

Reply:
OK, got it.